pddlgym/core.py

```diff
@@ -171,31 +203,80 @@ def _apply_effects(state, lifted_effects, assignments):
     new_literals = set(state.literals)
     determinized_lifted_effects = []
     # Handle probabilistic effects.
+    probabilistic_lifted_effects = []
     for lifted_effect in lifted_effects:
         if isinstance(lifted_effect, ProbabilisticEffect):
-            chosen_effect = lifted_effect.sample()
-            if chosen_effect == "NOCHANGE":
-                continue
-            if isinstance(chosen_effect, LiteralConjunction):
-                for lit in chosen_effect.literals:
-                    determinized_lifted_effects.append(lit)
+            effect_outcomes = lifted_effect.literals
+            cur_probabilistic_lifted_effects = []
+            if get_all_transitions:
+                lifted_effects_list = cur_probabilistic_lifted_effects
             else:
-                determinized_lifted_effects.append(chosen_effect)
+                lifted_effects_list = determinized_lifted_effects
+            sampled_effect = lifted_effect.sample()
+
+            effects_to_process = [
+                sampled_effect] if not get_all_transitions else effect_outcomes
+
+            for chosen_effect in effects_to_process:
+                if isinstance(chosen_effect, LiteralConjunction):
+                    for lit in chosen_effect.literals:
+                        lifted_effects_list.append(lit)
+                else:
+                    lifted_effects_list.append(chosen_effect)
+
+            if get_all_transitions:
+                probabilistic_lifted_effects.append(
+                    cur_probabilistic_lifted_effects)
         else:
             determinized_lifted_effects.append(lifted_effect)
 
-    for lifted_effect in determinized_lifted_effects:
-        effect = ground_literal(lifted_effect, assignments)
-        # Negative effect
-        if effect.is_anti:
-            literal = effect.inverted_anti
-            if literal in new_literals:
-                new_literals.remove(literal)
-    for lifted_effect in determinized_lifted_effects:
-        effect = ground_literal(lifted_effect, assignments)
-        if not effect.is_anti:
-            new_literals.add(effect)
-    return state.with_literals(new_literals)
+    # TODO
+    # - Refactor code after finished
+    states = []
+    if not get_all_transitions:
+        for lifted_effect in determinized_lifted_effects:
+            if lifted_effect == "NOCHANGE":
+                continue
+            effect = ground_literal(lifted_effect, assignments)
+            # Negative effect
+            if effect.is_anti:
+                literal = effect.inverted_anti
+                if literal in new_literals:
+                    new_literals.remove(literal)
+        for lifted_effect in determinized_lifted_effects:
+            if lifted_effect == "NOCHANGE":
+                continue
+            effect = ground_literal(lifted_effect, assignments)
+            if not effect.is_anti:
+                new_literals.add(effect)
```

Comment:
Is there any reason for these two for loops exist instead of just one with the if statements combined?

Reply:
Good question! Yes, this was something we did specifically to ensure that negative effects were processed before positive effects. We should make sure to keep it that way!